File: google_reference/app/mal_service/src/main/java/com/cltv/mal/model/pvr/Recording.java
package com.cltv.mal.model.pvr;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;

import com.cltv.mal.model.entities.TvChannel;
import com.cltv.mal.model.entities.TvEvent;

import java.util.Date;

public class Recording implements Parcelable {
    int id;
    String name;
    long duration;
    long recordingDate;
    String image;
    String videoUrl;
    TvChannel tvChannel;
    TvEvent tvEvent;
    long recordingStartTime;
    long recordingEndTime;
    String shortDescription;

    boolean isEventLocked = false;

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public long getDuration() {
        return duration;
    }

    public void setDuration(long duration) {
        this.duration = duration;
    }

    public long getRecordingDate() {
        return recordingDate;
    }

    public void setRecordingDate(long recordingDate) {
        this.recordingDate = recordingDate;
    }

    public String getImage() {
        return image;
    }

    public void setImage(String image) {
        this.image = image;
    }

    public String getVideoUrl() {
        return videoUrl;
    }

    public void setVideoUrl(String videoUrl) {
        this.videoUrl = videoUrl;
    }

    public TvChannel getTvChannel() {
        return tvChannel;
    }

    public void setTvChannel(TvChannel tvChannel) {
        this.tvChannel = tvChannel;
    }

    public TvEvent getTvEvent() {
        return tvEvent;
    }

    public void setTvEvent(TvEvent tvEvent) {
        this.tvEvent = tvEvent;
    }

    public long getRecordingStartTime() {
        return recordingStartTime;
    }

    public void setRecordingStartTime(long recordingStartTime) {
        this.recordingStartTime = recordingStartTime;
    }

    public long getRecordingEndTime() {
        return recordingEndTime;
    }

    public void setRecordingEndTime(long recordingEndTime) {
        this.recordingEndTime = recordingEndTime;
    }

    public String getShortDescription() {
        return shortDescription;
    }

    public void setShortDescription(String shortDescription) {
        this.shortDescription = shortDescription;
    }

    public boolean isInProgress() {
        return duration == 0;
    }

    public boolean isEventLocked() {
        return isEventLocked;
    }

    public void setEventLocked(boolean eventLocked) {
        isEventLocked = eventLocked;
    }

    public Recording(int id, String name, long duration, long recordingDate, String image, String videoUrl, TvChannel tvChannel, TvEvent tvEvent, long recordingStartTime, long recordingEndTime, String shortDescription) {
        this.id = id;
        this.name = name;
        this.duration = duration;
        this.recordingDate = recordingDate;
        this.image = image;
        this.videoUrl = videoUrl;
        this.tvChannel = tvChannel;
        this.tvEvent = tvEvent;
        this.recordingStartTime = recordingStartTime;
        this.recordingEndTime = recordingEndTime;
        this.shortDescription = shortDescription;
    }

    protected Recording(Parcel parcel) {
        id = parcel.readInt();
        name = parcel.readString();
        duration = parcel.readLong();
        recordingDate = parcel.readLong();
        image = parcel.readString();
        videoUrl = parcel.readString();
        tvChannel = parcel.readParcelable(TvChannel.class.getClassLoader());
        tvEvent = parcel.readParcelable(TvEvent.class.getClassLoader());
        recordingStartTime = parcel.readLong();
        recordingEndTime = parcel.readLong();
        shortDescription = parcel.readString();
        isEventLocked = parcel.readByte() != 0;

    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        dest.writeInt(id);
        dest.writeString(name);
        dest.writeLong(duration);
        dest.writeLong(recordingDate);
        dest.writeString(image);
        dest.writeString(videoUrl);
        dest.writeParcelable(tvChannel, 0);
        dest.writeParcelable(tvEvent, 0);
        dest.writeLong(recordingStartTime);
        dest.writeLong(recordingEndTime);
        dest.writeString(shortDescription);
        dest.writeByte((byte) (isEventLocked ? 1 : 0));
    }

    public static final Creator<Recording> CREATOR = new Creator<Recording>() {
        @Override
        public Recording createFromParcel(Parcel in) {
            return new Recording(in);
        }

        @Override
        public Recording[] newArray(int size) {
            return new Recording[size];
        }
    };

    @NonNull
    @Override
    public String toString() {
        return "Recording = [id = " + id + ", name = " + name + ", duration = " + duration + ", startTime = " + new Date(recordingStartTime) + ", endTime = " + new Date(recordingEndTime) + ", imagePath = " + image + ", uri = " + videoUrl + ", tv channel = " + tvChannel.name + ", event = " + tvEvent.name;
    }

    public static String createRecordingTimeInfo(long recordingDate, long recordingEndTime) {
        if (recordingDate != 0 && recordingEndTime != 0) {
            Date startTime = new Date(recordingDate);
            Date endTime = new Date(recordingEndTime);

            long totalSeconds = (endTime.getTime() - startTime.getTime()) / 1000;
            long hours = (totalSeconds / 3600) % 24;
            long minutes = (totalSeconds / 60) % 60;
            long seconds = totalSeconds % 60;
            if (totalSeconds < 0) {
                hours = 0;
                minutes = 0;
                seconds = 0;
            }
            if (hours > 0) {
                return (hours < 10 ? "0" + hours + " h" : hours + " h") + " " +
                        (minutes < 10 ? "0" + minutes + " min" : minutes + " min") + " " +
                        (seconds < 10 ? "0" + seconds + " sec" : seconds + " sec");
            } else if (minutes > 0) {
                return (minutes < 10 ? "0" + minutes + " min" : minutes + " min") + " " +
                        (seconds < 10 ? "0" + seconds + " sec" : seconds + " sec");
            } else {
                return seconds < 10 ? "0" + seconds + " sec" : seconds + " sec";
            }
        }
        return "No start time defined";
    }


    public static int calculateCurrentProgress(long currentTimeData, Long startTime, Long endTime) {
        if (startTime == null || endTime == null) {
            return 0;
        }
        long currentTime = currentTimeData;
        long currentMinute = (currentTime - startTime);
        long duration = (endTime - startTime);
        int progress = 0;
        if (duration > 0) {
            progress = (int) ((currentMinute * 100) / duration);
        }
        // Progress value
        return progress;
    }

}

File: google_reference/app/platform/src/main/java/com/iwedia/cltv/platform/model/recording/Recording.kt
package com.iwedia.cltv.platform.model.recording

import com.iwedia.cltv.platform.model.TvChannel
import com.iwedia.cltv.platform.model.TvEvent
import com.iwedia.cltv.platform.model.player.PlayableItem
import java.util.*

class Recording (
    var id : Int = -1,
    var name : String = "",
    var duration : Long = 0,
    var recordingDate: Long = 0,
    var image : String,
    var videoUrl : String,
    var tvChannel: TvChannel?= null,
    var tvEvent: TvEvent? = null,
    var recordingStartTime : Long,
    var recordingEndTime : Long,
    var shortDescription : String,
    var longDescription : String = "",
    var contentRating: String,
    var isEventLocked : Boolean = false
) : PlayableItem
{
    var recordedEvent: TvEvent ?= null
    var data: Any?= null
    override fun toString(): String {
        return "Recording = [id = $id, name = $name," +
                "duration = ${duration}, startTime = ${Date(recordingStartTime)}" +
                ", endTime = ${Date(recordingEndTime)}, imagePath = $image," +
                " uri = $videoUrl, tv channel = ${tvChannel?.name}, event = ${tvEvent?.name}"
    }

    fun isInProgress(): Boolean {
        return duration == 0L
    }

    companion object {

        /**
         * Creates tv event time info (HH:mm - HH:mm)
         *
         * @param recordingDate recording item date
         * @param recordingEndTime recording item end time
         * @return tv event time info
         */
        fun createRecordingTimeInfo(recordingDate: Long, recordingEndTime: Long): String? {
            return if (recordingDate != null && recordingEndTime != null) {
                var startTime = Date(recordingDate)
                var endTime = Date(recordingEndTime)

                if (recordingDate != null) {
                    startTime = Date(recordingDate)
                }
                if (recordingEndTime!! != null) {
                    endTime = Date(recordingEndTime)
                }
                var totalSeconds = (endTime.time - startTime.time) / 1000
                var hours = totalSeconds / 3600 % 24
                var minutes = totalSeconds / 60 % 60
                var seconds = totalSeconds % 60
                if (totalSeconds < 0) {
                    hours = 0
                    minutes = 0
                    seconds = 0
                }
                return if (hours > 0) {
                    (if (hours < 10) "0$hours h" else "$hours h").toString() + " " + (if (minutes < 10) "0$minutes min" else "$minutes min") + " " + if (seconds < 10) "0$seconds sec" else "$seconds sec"
                } else if (minutes > 0) {
                    (if (minutes < 10) "0$minutes min" else "$minutes min").toString() + " " + if (seconds < 10) "0$seconds sec" else "$seconds sec"
                } else {
                    if (seconds < 10) "0$seconds sec" else "$seconds sec"
                }
            } else {
                "No start time defined"
            }
        }

        /**
         * Calculate progress bar value
         *
         * @param currentTimeData current time
         * @param startTime item start time
         * @param endTime item end time
         * @return progress percentage
         */
        fun calculateCurrentProgress(currentTimeData: Long, startTime: Long, endTime: Long): Int {
            if (startTime == null || endTime == null) {
                return 0
            }
            val currentTime = currentTimeData
            val currentMinute = (currentTime - startTime)
            val duration = (endTime!! - startTime)

            var progress = 0
            if (duration > 0) {
                progress = (currentMinute * 100 / duration).toInt()
            }

            //Progress value
            return progress.toInt()
        }
    }
}
